### PR TITLE
DEX-737 Checker

### PR DESCRIPTION
**Pull request by DenisSerebryansky**
Checker for testing DEX state after update via Waves DEX CLI `check-server` command

Concept:
  * There are two main components:
    1. SuperConnector, is used for connection to REST/gRPC/... APIs of DEX/Node. Can be easily enriched by another connectors. Currently it includes few connectors: `DexRestConnector`, `NodeRestConnector` and `DexExtensionGrpcConnector`;
    2. Checker, contains logic of DEX state testing by means of superConnector methods and own ones
 * Creation of the SuperConnector and checks are written in fail-fast, pretty FP style;
 * All REST methods of connectors are basically the same as in IT module. Main difference is that in this tool we don't parse received Jsons into domain entities and use these Jsons as is;

This tool was tested with the docker containers from IT module

### CODE REVIEW

File: dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala [hunk 1]
```diff
@@ -176,6 +201,26 @@ object WavesDexCli {
...
+                    s"""
+                      |Passed arguments:
+                      |  Waves Node REST API  : ${args.nodeRestApi}
+                      |  Expected DEX version : ${args.version}
+                      |  DEX config path      : ${args.dexConfigPath}
+                   """.stripMargin
+                  )
+                  superConnector <- SuperConnector.create(args.dexConfigPath, args.nodeRestApi)
+                  checkResult    <- Checker(superConnector).checkState(args.version)
+                  _              <- tool.lift { superConnector.close() }
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala:218]: Sadly this won't execute if the checks fail :) Could you add TODO? I hope we will solve this in the future

File: dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  private def denormalizeWavesBalance(value: Long): Double = denormalize(value, 8)
+
+  private def printOrder(amountAssetName: String, priceAssetName: String)(order: Order): String =
+    s"${order.orderType} ${denormalize(order.amount)} $amountAssetName @ ${denormalize(order.price)} $priceAssetName, id = ${order.id()}"
+
+  private def getAmountAndPriceAssetsInfo(f: AssetInfo, s: AssetInfo): (AssetInfo, AssetInfo) =
+    if (f.asset.compatId < s.asset.compatId) s -> f else f -> s
+
+  private def checkVersion(version: String): CheckResult[Unit] = dexRest.swaggerRequest.flatMap { response =>
+    val parsedVersion = (response \ "info" \ "version").get.as[String]
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala:50]: 👍

File: dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala:218]: It does since SuperConnector is `AutoClosable`

Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala:218]: How? `AutoClosable` is an interface to work with Java's `try-with-resources` syntax

Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala:218]: idk its a magic, but works
Try it out!

Comment (vsuharnikov) [reply dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala:218]:
```scala
class Resource extends AutoCloseable {
  override def close(): Unit = println("resource closed")
}

println("Start")
for {
  _ <- Either.cond(test = true, (), "1")
  r <- Right(new Resource)
  _ <- Either.cond(test = false, (), "2")
  _ <- Right(r.close())
} yield ()
println("End")
```

prints in Ammonite:
```
defined class Resource
Start
res1: scala.util.Either[String,Unit] = Left(2)
End
```

File: dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      _                       <- logPassed
+
+      _                            <- logCheck("3. First test asset")
+      (wuJIoInfo, firstAssetNotes) <- checkTestAsset(balance, firstTestAssetName)
+      _                            <- logPassed
+
+      _                              <- logCheck("4. Second test asset")
+      (mbIJIoInfo, secondAssetNotes) <- checkTestAsset(balance, secondTestAssetName)
+      _                              <- logPassed
+
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    }
+  }
+
+  def checkState(version: String): ErrorOr[String] =
+    for {
+      _ <- log("\nChecking\n")
+
+      _ <- logCheck("1. DEX version")
+      _ <- checkVersion(version)
+      _ <- logPassed
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala]: :+1: Done

File: dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+      _                            <- logCheck("3. First test asset")
+      (wuJIoInfo, firstAssetNotes) <- checkTestAsset(balance, firstTestAssetName)
+      _                            <- logPassed
+
+      _                              <- logCheck("4. Second test asset")
+      (mbIJIoInfo, secondAssetNotes) <- checkTestAsset(balance, secondTestAssetName)
+      _                              <- logPassed
+
+      _                       <- logCheck("5. Order placement")
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/tool/Checker.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala:219]: gRPC client closed anyway, we can inspect this issue in another task

File: dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala [hunk 6]
```diff
@@ -176,6 +201,26 @@ object WavesDexCli {
...
+                } yield checkResult
+              ) match {
+                case Right(diagnosticNotes) => println(s"$diagnosticNotes\nCongratulations! All checks passed!")
+                case Left(error)            => println(error)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/WavesDexCli.scala]: Done

**Pull request merged**